### PR TITLE
Changes to support PR #219

### PR DESCRIPTION
**Pull request by celiot-IDM**
This PR contains changes to the way sheets are loaded and validated. This work was prompted by problems uncovered through PR #219.

The only visible change from this PR should be a lot more trace comments while sheets are being loaded.

image

### CODE REVIEW

**Review by MeWu-IDM — CHANGES_REQUESTED** [review]
some issues Need to be addressed:
(1) For optional columns, if the type is wrong and  (e$convertType = False), we should ignore that and just give a warning. Actual behavior :https://github.com/InstituteforDiseaseModeling/PACE-HRH/blob/charles/pr219/pacehrh/R/pace_table_tools.R#L28-L32 will result in InitializeScenarios to NULL and not able to run
(2) For required columns, conversion should be strict and not silently coerced to NAs (otherwise user may not realized they made a mistake until see weird results) 
(3) For FTEratio if we decide to make it optional, we should remove unused refeneces from the code to keep things clean 
(4) Discussed with @rhanIDM and she will be fixing the input_model_demo.xlsx before we merge so that the spreadsheet is in sync with the code

**Comment by celiot-IDM** [comment]
> (1) For optional columns, if the type is wrong and (e$convertType = False), we should ignore that and just give a warning. Actual behavior :https://github.com/InstituteforDiseaseModeling/PACE-HRH/blob/charles/pr219/pacehrh/R/pace_table_tools.R#L28-L32 will result in InitializeScenarios to NULL and not able to run

I don't think this would be the appropriate behavior.

The idea of optional columns is that the columns are required internally, but will be defaulted if not provided by the user in the spreadsheet. convertType == TRUE is a convenience: the system will try to make sense of columns even if they're not provided in the spreadsheet with the right type. What you're proposing is that a column provided by the user with the wrong type would get all the way into the system with the wrong type preserved.

Note: It creates a different type of chaos if optional columns with the wrong type are silently dropped. Then the system has to check everywhere for the existence of the column data it needs. The point of the design is to ensure that once the tables have been loaded, the system can assume they're valid in format.

Final note: one of the reasons for all these shenanigans is that read_xlsx() returns empty columns as type "logical". This makes sense from the import perspective ("I don't have any information on which to guess the column type, so I'll assume a type that can be safely converted to any other type"). But once in the system, it's important that columns have the right types.

**Comment by celiot-IDM** [comment]
> (2) For required columns, conversion should be strict and not silently coerced to NAs (otherwise user may not realized they made a mistake until see weird results)

So to be clear ... you're suggesting that the system stop instead? (The system isn't actually silent on this, right? It does give an "NAs introduced by coercion" warning? If that's not sufficient indication something isn't right, then stopping is the only remaining option.)

**Comment by celiot-IDM** [comment]
> (3) For FTEratio if we decide to make it optional, we should remove unused refeneces from the code to keep things clean

I agree. Let me know whether or not we want to rip out FTEratio and I'll do the deed.

**Comment by MeWu-IDM** [comment]
> > (1) For optional columns, if the type is wrong and (e$convertType = False), we should ignore that and just give a warning. Actual behavior :https://github.com/InstituteforDiseaseModeling/PACE-HRH/blob/charles/pr219/pacehrh/R/pace_table_tools.R#L28-L32 will result in InitializeScenarios to NULL and not able to run
> 
> I don't think this would be the appropriate behavior.
> 
> The idea of optional columns is that the columns are required internally, but will be defaulted if not provided by the user in the spreadsheet. convertType == TRUE is a convenience: the system will try to make sense of columns even if they're not provided in the spreadsheet with the right type. What you're proposing is that a column provided by the user with the wrong type would get all the way into the system with the wrong type preserved.
> 
My understanding is that optional columns are not used by the model (is it not true??) , but if they are required then they should be provided with default values regardless of type error or not. In this example, if user set Delivery_model in the scenario tab to 1, 2, 3 (loaded as double type), they would get a warning "NAs introduced by coercion" which does not help them to understand why it does not work. At the very least the warning message should be self-explanatory.

**Comment by MeWu-IDM** [comment]
> > (2) For required columns, conversion should be strict and not silently coerced to NAs (otherwise user may not realized they made a mistake until see weird results)
> 
> So to be clear ... you're suggesting that the system stop instead? (The system isn't actually silent on this, right? It does give an "NAs introduced by coercion" warning? If that's not sufficient indication something isn't right, then stopping is the only remaining option.)

I think we should stop the processing, if user had a typo on offset month and it casued the offset to be set to NA (NAs introduced by coercion warning) they may not be aware that this is something serious (which will likely cause the results to be unexpected from their input) @rhanIDM or @BHagedorn-IDM can chime in their opinion for the expected behavior.

**Comment by celiot-IDM** [comment]
> > > (1) For optional columns, if the type is wrong and (e$convertType = False), we should ignore that and just give a warning. Actual behavior :https://github.com/InstituteforDiseaseModeling/PACE-HRH/blob/charles/pr219/pacehrh/R/pace_table_tools.R#L28-L32 will result in InitializeScenarios to NULL and not able to run
> > 
> > 
> > I don't think this would be the appropriate behavior.
> > The idea of optional columns is that the columns are required internally, but will be defaulted if not provided by the user in the spreadsheet. convertType == TRUE is a convenience: the system will try to make sense of columns even if they're not provided in the spreadsheet with the right type. What you're proposing is that a column provided by the user with the wrong type would get all the way into the system with the wrong type preserved.
> 
> My understanding is that optional columns are not used by the model (is it not true??) , but if they are required then they should be provided with default values regardless of type error or not. In this example, if user set Delivery_model in the scenario tab to 1, 2, 3 (loaded as double type), they would not be able to initialize scenarios as it returns NULL which does not help them to understand why it does not work. At the very least the warning message should be self-explanatory.

Two issues:
- The simple issue: yes, we can implement a more informative experience when we detect irreconcilable type mismatches.
- The general issue relates to why irreconcilable type mismatches happen and the nature of "optional" columns. The required/optional setup is there to (1) support a degree of backwards compatibility; and (2) to support a future where we have optional *features*. Re (1): We've already released a v1, so we need to recognize a new class of user problems arising from using old spreadsheets. The standard deprecation practice is to tolerate old features (or in our case, columns that no longer trigger any processing) for a while before finally removing them. Re (2): This is the more important case. We've already started to introduce flags to control how much modelling is done (annual, monthly, etc). Cadre processing goes a step further: to do cadre processing, the system needs different data. But if the user never does cadre processing, that data it is truly optional.

**Comment by BHagedorn-IDM** [comment]
> > (1) For optional columns, if the type is wrong and (e$convertType = False), we should ignore that and just give a warning. Actual behavior :https://github.com/InstituteforDiseaseModeling/PACE-HRH/blob/charles/pr219/pacehrh/R/pace_table_tools.R#L28-L32 will result in InitializeScenarios to NULL and not able to run
> 
> I don't think this would be the appropriate behavior.
> 
> The idea of optional columns is that the columns are required internally, but will be defaulted if not provided by the user in the spreadsheet. convertType == TRUE is a convenience: the system will try to make sense of columns even if they're not provided in the spreadsheet with the right type. What you're proposing is that a column provided by the user with the wrong type would get all the way into the system with the wrong type preserved.
> 
> Note: It creates a different type of chaos if optional columns with the wrong type are silently dropped. Then the system has to check everywhere for the existence of the column data it needs. The point of the design is to ensure that once the tables have been loaded, the system can assume they're valid in format.
> 
> Final note: one of the reasons for all these shenanigans is that read_xlsx() returns empty columns as type "logical". This makes sense from the import perspective ("I don't have any information on which to guess the column type, so I'll assume a type that can be safely converted to any other type"). But once in the system, it's important that columns have the right types.

Chiming in on desired behavior, I think we should stop the model if an optional column has an incorrect format, so that the error doesn't silently get propagated through.  I will also set up the format requirements in the Excel sheet to reinforce this. The user could always remove those, but it would have to be an intentional choice then, so we shouldn't see this behavior occurring.

**Comment by BHagedorn-IDM** [comment]
> > (3) For FTEratio if we decide to make it optional, we should remove unused refeneces from the code to keep things clean
> 
> I agree. Let me know whether or not we want to rip out FTEratio and I'll do the deed.

Yes, let's remove it fully.

**Comment by celiot-IDM** [comment]
> Yes, let's remove it fully.

Will do!

**Comment by celiot-IDM** [comment]
> I think we should stop the model if an optional column has an incorrect format, so that the error doesn't silently get propagated through.

Cool. Working on it!

**Comment by celiot-IDM** [comment]
PS As fiddly as this work is, I'm glad we're doing it.

**Comment by celiot-IDM** [comment]
I still have to remove the FTEratio processing.

File: pacehrh/R/pace_task_manage.R [hunk 1]
```diff
@@ -1,21 +1,3 @@
-.taskColumnNames <- c(
-  "Indicator",
-  "CommonName",
-  "ClinicalOrNon",
-  "ClinicalCat",
-  "ServiceCat",
-  "RelevantPop",
-  "Geography",
```
Comment (MeWu-IDM) [inline pacehrh/R/pace_task_manage.R:8]: this "Geography" column is not in either optional or required but this column is still being checked by lint tool, we should remove that reference if it is not needed
https://github.com/InstituteforDiseaseModeling/PACE-HRH/blob/10ba0ddbf78d4495fd4302c0b9ccfa5d6d4d2b36/pacehrh/R/pace_lint.R#L298

File: pacehrh/R/pace_scenarios_manage.R [hunk 2]
```diff
@@ -1,41 +1,3 @@
...
-    "o_PopGrowth",
-    "o_Fertility_decr",
-    "o_MHIVTB_decr",
-    "o_ChildDis_decr",
-    "sheet_TaskValues",
-    "sheet_PopValues",
-    "sheet_SeasonalityCurves",
-    "sheet_Cadre",
-    "DeliveryModel",
-    "Geography_dontedit"
```
Comment (MeWu-IDM) [inline pacehrh/R/pace_scenarios_manage.R:16]: this column "Geography_dontedit" is still being used in https://github.com/InstituteforDiseaseModeling/PACE-HRH/blob/10ba0ddbf78d4495fd4302c0b9ccfa5d6d4d2b36/pacehrh/R/pace_post_process.R#L209 but this column is removed the demo scenario sheet

File: pacehrh/R/pace_task_manage.R (hunk 1 above)
Comment (celiot-IDM) [reply pacehrh/R/pace_task_manage.R:8]: Sorted!

File: pacehrh/R/pace_scenarios_manage.R (hunk 2 above)
Comment (celiot-IDM) [reply pacehrh/R/pace_scenarios_manage.R:16]: Also sorted.

**Review by MeWu-IDM — APPROVED** [review]

**Pull request merged**